sickcodes/Docker-OSX
# Custom Serial Numbers (iMessage/iCloud security analysis)

## How to test

```bash
apt update -y && apt install git curl wget vim xvfb scrot build-essential sshpass -y \
    && git clone https://github.com/stolk/imcat.git \
    && cd imcat \
    && make \
    && sudo cp imcat /usr/bin/imcat \
    && touch /usr/bin/scrotcat \
    && tee -a /usr/bin/scrotcat <<< '/usr/bin/imcat <(scrot -o /dev/stdout)' \
    && chmod +x /usr/bin/scrotcat \
&& BRANCH=custom-identity \
&& CACHE= \
&& export DISPLAY=:99 \
&& cd ~ \
    && ( nohup Xvfb :99 -screen 0 1920x1080x16 & ) \
    && echo 1 | tee /sys/module/kvm/parameters/ignore_msrs \
    ; rm -rf Docker-OSX \
    ; cd ~ \
    && git clone https://github.com/sickcodes/Docker-OSX.git \
    && cd Docker-OSX \
    && git checkout -f "${BRANCH}" \
    && docker build --build-arg RANKMIRRORS=true  --build-arg MIRROR_COUNTRY=US ${CACHE} -t docker-osx:latest . \
    && docker tag docker-osx:latest sickcodes/docker-osx:latest \
    && docker build --build-arg RANKMIRRORS=true  --build-arg MIRROR_COUNTRY=US ${CACHE} -t docker-osx:naked -f Dockerfile.naked . \
    && docker tag docker-osx:naked sickcodes/docker-osx:naked \
    && docker build --build-arg RANKMIRRORS=true  --build-arg MIRROR_COUNTRY=US ${CACHE} -t docker-osx:auto -f Dockerfile.auto . \
    && docker tag docker-osx:auto sickcodes/docker-osx:auto

```

After that all of these should work.

For testing headless, you can take screenshots in the terminal now using `scrotcat`.

To test using headless, open a separate terminal with:
```
export DISPLAY=:99
scrotcat

while true; do scrotcat; sleep 1; done
```

All the command below are examples of different ways to set serial numbers.
```bash
cd ~
wget https://images2.sick.codes/mac_hdd_ng_auto.img

cp mac_hdd_ng_auto.img mac_hdd_ng_testing.img

cd ~

touch C02TW0WAHX87.env
touch ./C02TW0WAHX87.qcow

# successful
docker run -it \
    --device /dev/kvm \
    -e "DISPLAY=${DISPLAY:-:0.0}" \
    -v /tmp/.X11-unix:/tmp/.X11-unix \
    -p 50922:10022 \
    -e NOPICKER=true \
    -e GENERATE_SPECIFIC=true \
    -e DEVICE_MODEL="iMacPro1,1" \
    -e SERIAL="C02TW0WAHX87" \
    -e BOARD_SERIAL="C027251024NJG36UE" \
    -e UUID="5CCB366D-9118-4C61-A00A-E5BAF3BED451" \
    -e MAC_ADDRESS="A8:5C:2C:9A:46:2F" \
    -e BOOTDISK=/bootdisk \
    -v "${PWD}/C02TW0WAHX87.qcow:/bootdisk" \
    -v "${PWD}/mac_hdd_ng_testing.img:/image" \
    sickcodes/docker-osx:naked

# sshpass -p alpine ssh user@localhost -p 50922 'ioreg -l | grep IOPlatformSerialNumber'

until [[ "$(sshpass -p alpine ssh user@localhost -p 50922 'ioreg -l | grep IOPlatformSerialNumber')" = '    |   "IOPlatformSerialNumber" = "C02TW0WAHX87"' ]]; do
    sleep 1
done

docker run -it \
    --device /dev/kvm \
    -v "${PWD}/mac_hdd_ng.img:/image" \
    -v /tmp/.X11-unix:/tmp/.X11-unix \
    -p 50922:10022 \
    -e NOPICKER=true \
    -e GENERATE_SPECIFIC=true \
    -e DEVICE_MODEL="iMacPro1,1" \
    -e SERIAL="C02TW0WAHX87" \
    -e BOARD_SERIAL="C027251024NJG36UE" \
    -e UUID="5CCB366D-9118-4C61-A00A-E5BAF3BED451" \
    -e MAC_ADDRESS="A8:5C:2C:9A:46:2F" \
    -e BOOTDISK=/bootdisk \
    -v "${PWD}/C02TW0WAHX87.qcow:/bootdisk" \
    -e OSX_COMMANDS='ioreg -l | grep IOPlatformSerialNumber' \
    sickcodes/docker-osx:auto

docker run -it \
    --device /dev/kvm \
    -e "DISPLAY=${DISPLAY:-:0.0}" \
    -v /tmp/.X11-unix:/tmp/.X11-unix \
    -p 50922:10022 \
    -e NOPICKER=true \
    -e GENERATE_UNIQUE=true \
    -e GENERATE_SPECIFIC=true \
    -e DEVICE_MODEL="iMacPro1,1" \
    -e BOOTDISK=/bootdisk \
    -v "${PWD}/C02TW0WAHX87.qcow:/bootdisk" \
    -v "${PWD}/C02TW0WAHX87.env:/env" \
    -v "${PWD}/mac_hdd_ng_testing.img:/image" \
    sickcodes/docker-osx:naked

until [[ "$(sshpass -p alpine ssh user@localhost -p 50922 'ioreg -l | grep IOPlatformSerialNumber')" = '    |   "IOPlatformSerialNumber" = "C02TW0WAHX87"' ]]; do
    sleep 1
done

stat C02TW0WAHX87.qcow
cat C02TW0WAHX87.env

# this image generates a brand new machine, persisting the env file.
# to use this machine later, switch it to specific and use the newly generated values in the env files
docker run -it \
    --device /dev/kvm \
    -v "${PWD}/mac_hdd_ng.img:/image" \
    -v /tmp/.X11-unix:/tmp/.X11-unix \
    -p 50922:10022 \
    -e NOPICKER=true \
    -e GENERATE_UNIQUE=true \
    -e DEVICE_MODEL="iMacPro1,1" \
    -e BOOTDISK=/bootdisk \
    -v "${PWD}/C02TW0WAHX87.env:/env" \
    -e OSX_COMMANDS='ioreg -l | grep IOPlatformSerialNumber' \
    sickcodes/docker-osx:auto


# this image aims to persist the qcow by adding these two lines
    # -e GENERATE_SPECIFIC=true \
    # -v "${PWD}/C02TW0WAHX87.qcow:/bootdisk" \
docker run -it \
    --device /dev/kvm \
    -v "${PWD}/mac_hdd_ng.img:/image" \
    -v /tmp/.X11-unix:/tmp/.X11-unix \
    -p 50922:10022 \
    -e NOPICKER=true \
    -e GENERATE_UNIQUE=true \
    -e GENERATE_SPECIFIC=true \
    -e DEVICE_MODEL="iMacPro1,1" \
    -e BOOTDISK=/bootdisk \
    -v "${PWD}/C02TW0WAHX87.qcow:/bootdisk" \
    -v "${PWD}/C02TW0WAHX87.env:/env" \
    -e OSX_COMMANDS='ioreg -l | grep IOPlatformSerialNumber' \
    sickcodes/docker-osx:auto

# a naked run using the testing image
# this tests that on each run it should use a random serial number set.
# no persistent env information
# DONT USE THIS IF YOU WANT TO HAVE ONE DEVICE
docker run -it \
    --device /dev/kvm \
    -e "DISPLAY=${DISPLAY:-:0.0}" \
    -v /tmp/.X11-unix:/tmp/.X11-unix \
    -p 50922:10022 \
    -e NOPICKER=true \
    -e GENERATE_UNIQUE=true \
    -e DEVICE_MODEL="iMacPro1,1" \
    -v "${PWD}/mac_hdd_ng_testing.img:/image" \
    sickcodes/docker-osx:naked


until [[ "$(sshpass -p alpine ssh user@localhost -p 50922 'ioreg -l | grep IOPlatformSerialNumber')" = '    |   "IOPlatformSerialNumber" = "xxxxxxxxxxxxx"' ]]; do
    sleep 1
done


docker run -it \
    --device /dev/kvm \
    -v "${PWD}/mac_hdd_ng.img:/image" \
    -v /tmp/.X11-unix:/tmp/.X11-unix \
    -p 50922:10022 \
    -e NOPICKER=true \
    -e GENERATE_UNIQUE=true \
    -e DEVICE_MODEL="iMacPro1,1" \
    -e OSX_COMMANDS='ioreg -l | grep IOPlatformSerialNumber' \
    sickcodes/docker-osx:auto


```
